currency/forexprovider/base/base_interface.go

```diff
@@ -17,29 +16,141 @@ type IFXProvider interface {
 	GetName() string
 	IsEnabled() bool
 	IsPrimaryProvider() bool
+	GetSupportedCurrencies() ([]string, error)
+}
+
+// FXHandler defines a full suite of FX data providers with failure backup with
+// unsupported currency shunt procedure
+type FXHandler struct {
+	Primary Provider
+	Support []Provider
+	mtx     sync.Mutex
+}
+
+// Provider defines a singular foreign exchange provider with its supported
+// currencies to cross reference request currencies and if not supported shunt
+// request traffic to and from other providers so that we can maintain full
+// currency list integration
+type Provider struct {
+	Provider            IFXProvider
+	SupportedCurrencies []string
+}
+
+// GetNewRate access rates by predetermined logic based on how a provider
+// handles requests
+func (p *Provider) GetNewRate(base string, currencies []string) (map[string]float64, error) {
+	if !p.Provider.IsEnabled() {
+		return nil, fmt.Errorf("provider %s is not enabled",
+			p.Provider.GetName())
+	}
+
+	switch p.Provider.GetName() {
+	case "ExchangeRates":
+		return p.Provider.GetRates(base, "") // Zero value to get all rates
+
+	default:
+		return p.Provider.GetRates(base, common.JoinStrings(currencies, ","))
+	}
+}
+
+// CheckCurrencies cross references supplied currencies with exchange supported
+// currencies, if there are any currencies not supported it returns a list
+// to pass on to the next provider
+func (p Provider) CheckCurrencies(currencies []string) []string {
+	var spillOver []string
+	for _, c := range currencies {
+		if !common.StringDataCompareUpper(p.SupportedCurrencies, c) {
+			spillOver = append(spillOver, c)
+		}
+	}
+	return spillOver
 }
 
 // GetCurrencyData returns currency data from enabled FX providers
-func (fxp IFXProviders) GetCurrencyData(baseCurrency, symbols string) (map[string]float64, error) {
-	for x := range fxp {
-		if fxp[x].IsPrimaryProvider() && fxp[x].IsEnabled() {
-			rates, err := fxp[x].GetRates(baseCurrency, symbols)
+func (f *FXHandler) GetCurrencyData(baseCurrency string, currencies []string) (map[string]float64, error) {
+	var fullRange []string
+	if !common.StringDataCompareUpper(currencies, baseCurrency) {
+		fullRange = append(currencies, baseCurrency)
+	} else {
+		fullRange = currencies
+	}
+
+	f.mtx.Lock()
+	defer f.mtx.Unlock()
+
+	if f.Primary.Provider == nil {
+		return nil, errors.New("primary foreign exchange provider details not set")
+	}
+
+	shunt := f.Primary.CheckCurrencies(fullRange)
```

Comment:
Is shunt the right term here? (I've just not heard of it)

Reply:
term push, shove or divert, I am from an electrical background. What term would you recommend. I can also have spillOver.